Add generic::map_repr + improve documentation of the module.

Centril opened the pull request:
Pokes back at @ExpHP ;)

I was a bit slow =P but here's the proposed `map_repr` and some documentation improvements.

ExpHP commented on core/src/generic.rs:
```diff
@@ -67,47 +77,71 @@
+    Mapper: FnOnce(Repr) -> Repr
+{
+    <Elt as Generic>::map_repr(src, mapper)
+}
```
help contribute to world peace; configure your editor to always leave a trailing newline

Centril replied:
Oh no; not again... I thought I fixed that once and for all :/

ExpHP commented on core/src/generic.rs:
```diff
@@ -67,47 +77,71 @@
-    /// From one type to another using a type with a compatible generic representation
-    fn convert_from<A>(a: A) -> Self
+    /// Convert an object to a another type provided that they have
+    /// a compatible representation type.
```
Maybe this should say "the same representation type."

ExpHP commented on core/src/generic.rs:
```diff
@@ -67,47 +77,71 @@
 
-/// Given an A, returns its generic Representation
-pub fn into_generic<A, Repr>(a: A) -> Repr
+/// Given a object of type `Src`, returns its generic representation `Repr`.
```
**an** object.  (or perhaps better, "a value")

ExpHP commented on core/src/generic.rs:
```diff
@@ -67,47 +77,71 @@
-/// Given an A, returns its generic Representation
-pub fn into_generic<A, Repr>(a: A) -> Repr
+/// Given a object of type `Src`, returns its generic representation `Repr`.
+pub fn into_generic<Src, Repr>(src: Src) -> Repr
```
I'm just noticing this now, but maybe Repr shouldn't be in the type parameter list here (since it's not necessary, and the type parameter list is public API).

ExpHP commented on core/src/generic.rs:
```diff
@@ -67,47 +77,71 @@
-pub fn convert_from<A, B, Repr>(a: A) -> B
+/// Converts one type `Src` into another type `Dst` assuming they have the same
+/// generic representation.
+pub fn convert_from<Src, Dst, Repr>(src: Src) -> Dst
```
(if we do want to eliminate unnecessary type params, there's this one too, and probably more in `labelled`)

ExpHP commented on core/src/generic.rs:
```diff
@@ -67,47 +77,71 @@
+    <Dst as Generic>::convert_from(src)
 }
+
+/// Maps an object of the element type `Elt` using a function on the
```
I find "element" confusing, it makes me think it is referring to an individual item in the HList.

ExpHP commented on core/src/generic.rs:
```diff
@@ -80,14 +80,14 @@ pub trait Generic {
     fn into(self) -> Self::Repr;
 
-    /// Convert an object's representation type `Repr` to the object's type.
+    /// Convert a value's representation type `Repr` to the values's type.
```
values's =P

ExpHP commented on core/src/generic.rs:
```diff
@@ -80,14 +80,14 @@ pub trait Generic {
 
-    /// Convert an object to a another type provided that they have
-    /// a compatible representation type.
+    /// Convert a value to an another type provided that they have
```
an another

ExpHP commented on core/src/labelled.rs:
```diff
@@ -1,15 +1,18 @@
 //!
-//! Asides from the main LabelledGeneric trait, this module holds helper methods that allow
-//! users to use LabelledGeneric without using universal function call syntax.
+//! Asides from the main `LabelledGeneric` trait, this module holds helper
```
just spotted this old typo "Asides from"

ExpHP commented on core/src/labelled.rs:
```diff
@@ -143,48 +148,51 @@ pub trait LabelledGeneric {
 
-/// Given a labelled generic Representation of an A, returns A
-pub fn from_labelled_generic<A, Repr>(gen: Repr) -> A
+/// Given a labelled generic representation of an `Dst`, returns `Dst`
```
an Dst

Centril commented:
@ExpHP Yeah no, the macros would take time I didn't have ;)

Fixed most of your nits..

Also, I noticed that there are deprecated functions in `generic.rs` and `labelled.rs`. Should we remove them?

I added `map_inter` which is more flexible in some cases (when the intermediate type is known).

Re. the extra type parameters, did you want to remove them for all the free functions?

ExpHP commented:
Something makes me uncomfortable about the Generic impl for HLists because it is inherently different from all of the others (it's the only time that Generic is ever recursive).  If anything I would've expected Generic for HLists to be the identity transform.  I think we should maybe hold back on that and discuss it

Centril commented:
@ExpHP I initially wrote `type Repr = Self;`. That also works for me. =) I'll wait for @lloydmeta's review first tho before changing.

ExpHP commented:
P.S. I have no idea what `intern` is supposed to be short for, but I'm also half asleep right now

Centril commented:
@ExpHP Intermediate ;) I was really trying to find a better name but it was really hard 😆 I'd love a better name.

Centril commented on core/src/generic.rs:
```diff
@@ -32,10 +32,20 @@
+///
+/// Any implementation of `Generic` must satisfy the following two laws:
+///
+/// 1. `forall x : Self. x == Generic::from(Generic::into(x))`
```
Yeah probably :)

Centril commented:
@lloydmeta So what are your thought on the `Generic` impl for `HCons`?

lloydmeta commented:
@Centril, the `Generic` impl looks fine to me. If anything at all, I was wondering about the `H: Generic` bound and whether it would make sense to provide `Generic` impls for primitive types 🤔

ExpHP commented:
Providing identity Generic impls for standard types seems like a big rabbit hole.

I just took a look at [what Haskell does](http://hackage.haskell.org/package/base-4.11.0.0/docs/GHC-Generics.html#t::-42-:) and I'm... not actually sure *what* they do.  None of the Generic impls have any constraints, including the ones most closely corresponding to this:

```haskell
instance Generic1 (f :*: g :: k -> *)
instance Generic ((f :*: g) p)
```

which seems to suggest that the transformation for `:*:` is either the identity transform or simply one that adds some extra wrapper types for metadata. But *all* of the instances are generated with DeriveGeneric (*including* the ones for `:*:`), and the `Rep` types aren't documented, so I don't have much to go on. :confused:

---

My preference still is to take no action until a very strong motivation for adding `impl Generic for HCons` arises (not just "because we can," but rather "because we *need to*"); this will give us a better understanding of why a certain definition is the right one.

Centril commented:
Your take no action argument makes sense to me; I'll remove the impls for now (tomorrow I mean).

So this is what `Rep ((f :*: g) p)` is defined as:
```haskell
type Rep ((f :*: g) p) =
  D1
    (MetaData ":*:" "GHC.Generics" "base" False)
    (C1
      (MetaCons ":*:" (InfixI RightAssociative 6) False)
      (S1
        (MetaSel
          (Nothing :: Maybe Symbol)
          NoSourceUnpackedness
          NoSourceStrictness
          DecidedLazy
        )
        (Rec0 (f p))
        :*:
        S1
          (MetaSel
            (Nothing :: Maybe Symbol)
            NoSourceUnpackedness
            NoSourceStrictness
            DecidedLazy
          )
          (Rec0 (g p))
       )
     )
```

lloydmeta commented:
> My preference still is to take no action until a very strong motivation for adding impl Generic for HCons arises (not just "because we can," but rather "because we need to"); this will give us a better understanding of why a certain definition is the right one.

Yeah, letting the need drive the implementation sounds like a solid idea :+1:

Centril commented:
Removed the Generic for HList impls now =) @ExpHP merge if you're comfortable with the changes now?

ExpHP commented:
:shipit:
